### PR TITLE
Cryptography

### PR DESCRIPTION
**Pull request by HarrisonBaxter**
Providing support for encryption and decryption without any dependencies by making use of underlying Java cryptographic libraries. Provide are four different flows to encrypt and decrypt bytestring to bytestring Using symmetric and asymmetric algorithms.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @HarrisonBaxter,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by HarrisonBaxter** [comment]
I have signed the CLA

**Comment by ennru** [comment]
I don't understand who is the author of this PR.
Isn't this the work @fmsbeekmans started off in #702 ?

**Comment by fmsbeekmans** [comment]
It is, we worked together on this.

**Comment by ennru** [comment]
Ok, so you may close #702 than.
Put the documentation in the Data Transformations section, please.
I'll have a more thorough look another day.

**Comment by ennru** [comment]
I realised this simple approach won't work.
If you e.g. stream the encrypted ByteStrings to a file and than read it again, you won't get the same bytes per ByteString and won't be able to decrypt. There needs to be some frameing so that the correct part of the byte stream can be used for decryption.
This is related to https://github.com/akka/akka/issues/24361

**Comment by fmsbeekmans** [comment]
Will try and make a failing test case for this over the weekend. We are testing arbitrarily chunked strings, not bytes.

**Comment by ennru** [comment]
To test this, you need to encrypt a couple of ByteStrings, concat those, split them differently and decrypt.

**Comment by fmsbeekmans** [comment]
Tested with bytestrings chunked per byte, still works. The encryption and decryption works on `Array[Byte]`, not on `String`.

File: cryptography/src/test/scala/akka/stream/alpakka/cryptography/scaladsl/CryptographicFlowSpec.scala [hunk 1]
```diff
@@ -41,11 +41,27 @@ class CryptographicFlowSpec extends WordSpec with Matchers with ScalaFutures wit
...
+        val keyGenerator = KeyGenerator.getInstance("AES")
+        val secretKey = keyGenerator.generateKey()
+
+        val byteString = ByteString("byte by byte")
+
+        val res = Source.single(byteString)
+          .mapConcat(bs => bs.toArray.toList)
+          .map(b => ByteString(b))
+          .via(symmetricEncryption(secretKey))
+          .via(symmetricDecryption(secretKey))
```
Comment (ennru) [inline cryptography/src/test/scala/akka/stream/alpakka/cryptography/scaladsl/CryptographicFlowSpec.scala]: Add a concat between the encryption and decryption and you'll see what I mean. If we - for example - send the encrypted data via a socket or http, we'll get other ByteStrings on the other side. We'd need some frameing to cut out the same ByteStrings for decryption.

Comment (fmsbeekmans) [reply cryptography/src/test/scala/akka/stream/alpakka/cryptography/scaladsl/CryptographicFlowSpec.scala]: Oooh you're right, will have a look before the weekend is over!

**Comment by fmsbeekmans** [comment]
Added a chunking stage to the symmetric encryption and decryption. Still have to figure out the chunk size based on the key.

Realized that for the asymmetric encryption it should encrypt or decrypt the complete data at once. Not sure whether it makes sense for a flow or not.

**Comment by fmsbeekmans** [comment]
Should be able to make some time to look into this again soon.

**Comment by git-josip** [comment]
@fmsbeekmans  What is the status of this PR? Do we know is it going to be merged anytime soon , so client side encryption for Amazon S3 can be used.

**Comment by fmsbeekmans** [comment]
Still quite a bit of work to be done sadly.

**Comment by git-josip** [comment]
@fmsbeekmans Thanks, appreciate your work!

**Comment by ennru** [comment]
This PR hasn't been touched for a very long time.
Please re-open or create a new to continue this work.

**Pull request closed without merging**